File: react-app/src/components/ShoppingList/forms/ShoppingListItem.js
import { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useParams } from 'react-router'
import { loadMeasurements } from '../../../store/items'
import { addEditShoppingListItem } from '../../../store/shoppingList'
import './ShoppingListForms.css'

export default function ShoppingListItem({item, setShowModal, hideMenu}) {
  const dispatch = useDispatch();
  const { id: shoppingListId } = useParams();
  const measurements = useSelector(state => state.items.measurements);

  const [loaded, setLoaded] = useState(false);
  const [measurementId, setMeasurementId] = useState("");
  const [quantity, setQuantity] = useState("");
  const [errors, setErrors] = useState("")

  useEffect(() => {
    dispatch(loadMeasurements())
    setLoaded(true)
  },[dispatch]);

  const handleSubmit = async (e) => {
    setErrors("")
    e.preventDefault();
    const shoppingListItem = {
      measurementId,
      quantity,
      shoppingListId,
      itemId: item.id,
      method: 'POST'
    }
    const res = await dispatch(addEditShoppingListItem(shoppingListItem))
    if (res.errors) {
      setErrors(res.errors)
    }
    else {
      setShowModal(false)
      hideMenu()
    }
  }

  if (!loaded) return null;
  return (
    <form className="shoppingListForm" onSubmit={handleSubmit}>
      {errors && errors.map(error => <li key={error}>{error}</li>)}
      {item.name}
      {item.category}
      <input type="number" value={quantity} onChange={e => setQuantity(e.target.value)} />
      <select value={measurementId} onChange={e => setMeasurementId(e.target.value) }>
        {measurements && measurements.map(measurement => <option
                                          value={measurement.id}
                                          key={measurement.id}>
                                            {measurement.unit}
                                          </option>)
                                          }
      </select>
      <button>Add To List</button>
    </form>
  )
};

File: app/models/shopping_list_item.py
from .db import db


class ShoppingListItem(db.Model):
    __tablename__ = "shopping_list_items"

    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey("users.id"))
    item_id = db.Column(db.Integer, db.ForeignKey("items.id"))
    in_cart = db.Column(db.Boolean, default=False)
    shopping_list_id = db.Column(db.Integer,
                                 db.ForeignKey("shopping_lists.id"))
    quantity = db.Column(db.Float, nullable=False)
    measurement_id = db.Column(db.Integer, db.ForeignKey("measurements.id"),
                               nullable=False)
    measurement = db.relationship("Measurement",
                                  back_populates="shopping_list_items")
    shopping_list = db.relationship("ShoppingList",
                                    back_populates="shopping_list_items")
    item = db.relationship("Item", back_populates="shopping_list_items")

    def to_dict(self):
        return {
            "id": self.id,
            "item": self.item.to_dict(),
            "shopping_list_id": self.shopping_list_id,
            "quantity": self.quantity,
            "measurement": self.measurement.to_dict()
        }
